skimage/filters/_gabor.py

```diff
@@ -72,30 +90,134 @@ def gabor_kernel(frequency, theta=0, bandwidth=1, sigma_x=None, sigma_y=None,
     >>> io.imshow(gk.real)  # doctest: +SKIP
     >>> io.show()           # doctest: +SKIP
     """
-    if sigma_x is None:
-        sigma_x = _sigma_prefactor(bandwidth) / frequency
-    if sigma_y is None:
-        sigma_y = _sigma_prefactor(bandwidth) / frequency
+    def __new__(cls, frequency, thetas=None, bandwidth=1, sigma=None,
+                sigma_y=None, n_stds=3, offset=None, ndim=2, **kwargs):
```

Comment:
I'm not sure that the complexity of a class is necessary. Why did you pick this design instead of extending the existing function?

Reply:
@jni I chose this design because it encourages reusing the existing gabor kernel. As of right now, the `filters.gabor` function creates a new gabor kernel every time you use it. This becomes very expensive if I, say, want to apply the same gabor filter to 1000 images.

While there is already a method of eliminating this expense by creating a gabor kernel and applying it to the images yourself, I would posit that most users would not consider this as they would immediately gravitate towards the `gabor` function, which is not obviously expensive from the documentation when applied many times. Furthermore, if one wanted to reproduce what the `gabor` function did, they would have to look at the source code and figure out exactly how we were able to apply the filter to an image.

Although I generally prefer functional design to object-oriented design, I think that this choice is justified for the following reasons:
- NumPy is already object-oriented. If we're representing the kernel as an `np.ndarray` anyways, it would be most useful to extend this interface to apply it as well.
- This tags along the filter function to the kernel instead of having to use one function to make the kernel and another to apply it (this would be the functional alternative). This is also a lot more intuitive.

Reply:
@kne42  I think the simplest option is to add to `filters.gabor` a `kernel` parameter that is the (bare) NumPy array, and avoid recomputing the kernel if this argument is given.

The NumPy community has moved away from inheriting from NumPy arrays because it raises a whole bunch of problems. An example: suppose someone comes up with a fancy new "Gabor++" method that works by creating a Gabor kernel, modifying in some weird way, then applying it. Any call to `np.copy(kernel)` will strip away the class and return a bare array. Then you've lost your way of applying the kernel.

The other issue is the "surprise" factor: this is a pattern that does not exist anywhere in skimage, or NumPy/SciPy for that matter. So now users have to wrap their heads around a whole new API for Gabor kernels. Or they run `g = filters.gabor_kernel(freq)` and then type `type(g)` and get a strange class that they've never seen.